### PR TITLE
Add tag API support 

### LINKED ISSUE #221: Add Tag support to SDK

### PR DESCRIPTION
**Pull request by sav007**
Introduce product tag API support
Refactor product API endpoints

@bgulanowski @davidmuzi @krisorr 

closes #221

### CODE REVIEW

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductService.java [hunk 1]
```diff
@@ -111,59 +112,63 @@
     Observable<List<Product>> getProducts(List<Long> productIds);
 
     /**
-     * Fetch the list of Products in a Collection using the sort order defined in the shop admin
+     * Fetch a page of collections
```
Comment (bgulanowski) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductService.java]: Normally, we should not mix functional changes with basic file reorganization changes. It makes it harder to tell what changes have consequences, and which are superficial.

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java [hunk 2]
```diff
@@ -166,60 +161,89 @@ public CancellableTask getProducts(final List<Long> productIds, final Callback<L
...
-            .compose(new UnwrapRetrofitBodyTransformer<ProductListings, List<Product>>())
-            .onErrorResumeNext(new BuyClientExceptionHandler<List<Product>>())
+            .compose(new UnwrapRetrofitBodyTransformer<ProductTagsWrapper, List<ProductTag>>())
+            .map(new Func1<List<ProductTag>, List<String>>() {
+                @Override
+                public List<String> call(final List<ProductTag> productTags) {
+                    if (productTags == null) {
+                        return null;
+                    } else {
+                        final List<String> tags = new ArrayList<>();
```
Comment (bgulanowski) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java]: I know it's trivial, but would be much nicer if this mapping was extracted to a separate function.

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java [hunk 3]
```diff
@@ -166,60 +161,89 @@ public CancellableTask getProducts(final List<Long> productIds, final Callback<L
...
         if (page < 1) {
             throw new IllegalArgumentException("page is a 1-based index, value cannot be less than 1");
         }
 
-        // All collection responses from the server are wrapped in a CollectionListings object which contains and array of collections
-        // For this call, we will clamp the size of the collection array returned to the page size
+        final String tagsQueryStr = tags != null && !tags.isEmpty() ? TextUtils.join(",", tags.toArray()) : null;
+
+        final String sortOrderStr;
+        if (collectionId != null) {
```
Comment (bgulanowski) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java]: This should be refactored to be `if/else if/else`, instead of using conditional operator.

Comment (bgulanowski) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java]: Actually, can you make `sortOrder` default to `Collection.SortOrder.COLLECTION_DEFAULT` (or `if (sortOrder == null) sortOrder = DEFAULT`), and initialize `sortOrderStr` to `null` (then won't need the `else` block). Then `if (collectionId != null)` is the only thing we need to check.

**Comment by bgulanowski** [comment]
A few minor suggestions, otherwise looks OK.

File: MobileBuy/buy/src/main/java/com/shopify/buy/model/ProductTag.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ *   THE SOFTWARE.
+ */
+package com.shopify.buy.model;
+
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * Represents tag information attached to the product
+ */
+public class ProductTag extends ShopifyObject {
```
Comment (davidmuzi) [inline MobileBuy/buy/src/main/java/com/shopify/buy/model/ProductTag.java]: The API now returns the `title`, not the `handle`.

But, more importantly, do we really need this class? All it does is wrap a String, and is inconstant with the product model which uses an array of `Strings` for its `tags` property.

Comment (sav007) [inline MobileBuy/buy/src/main/java/com/shopify/buy/model/ProductTag.java]: This class not only wraps String it extends the Shopify object that has id field. So we have id, handle. And this class is used only internally, API returns List of Strings not List of ProductTags. Take a look at the response from server API

`
{
"tags": [3]
0:  {
"id": 12562217347
"title": "Vintage"
}-
1:  {
"id": 12562217283
"title": "summer"
}-
2:  {
"id": 12562217219
"title": "cotton"
}-
-
}
`

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java (hunk 2 above)
Comment (sav007) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java]: Done.

**Comment by krisorr** [comment]
Is there a related ticket?

**Comment by krisorr** [comment]
closes #221

File: MobileBuy/buy/src/androidTest/java/com/shopify/buy/service/BuyTest.java [hunk 5]
```diff
@@ -593,6 +593,48 @@ public void failure(BuyClientError error) {
...
 
+    @Test
+    public void testGetProductTags() throws Exception {
+        final CountDownLatch latch = new CountDownLatch(1);
+        buyClient.getProductTags(1, new Callback<List<String>>() {
+            @Override
+            public void success(List<String> response) {
+                assertNotNull(response);
+                latch.countDown();
+                // well we can't really test response for some values, as we can't control from client list of tags
```
Comment (krisorr) [inline MobileBuy/buy/src/androidTest/java/com/shopify/buy/service/BuyTest.java]: We control the test shop - why can't we check that we got the correct tags here?

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/BuyClientDefault.java [hunk 6]
```diff
@@ -496,32 +497,32 @@ public CancellableTask getProducts(List<Long> productIds, Callback<List<Product>
...
-    public Observable<List<Product>> getProducts(int page, Long collectionId, SortOrder sortOrder) {
-        return productService.getProducts(page, collectionId, sortOrder);
+    public Observable<List<String>> getProductTags(int page) {
+        return productService.getProductTags(page);
     }
 
     @Override
-    public CancellableTask getCollections(int page, Callback<List<Collection>> callback) {
-        return productService.getCollections(page, callback);
+    public CancellableTask getProducts(int page, Long collectionId, Set<String> tags, SortOrder sortOrder, Callback<List<Product>> callback) {
```
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/BuyClientDefault.java]: not a huge fan of these many arguments in an API function.  If we get more we should switch to a fluent parameter builder.

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductService.java (hunk 1 above)
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductService.java]: agreed, this is a bit hard to parse

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/BuyClientDefault.java [hunk 7]
```diff
@@ -496,32 +497,32 @@ public CancellableTask getProducts(List<Long> productIds, Callback<List<Product>
...
+        return productService.getProducts(page, collectionId, tags, sortOrder, callback);
     }
 
     @Override
-    public Observable<List<Collection>> getCollections(int page) {
-        return productService.getCollections(page);
+    public Observable<List<Product>> getProducts(int page, Long collectionId, Set<String> tags, SortOrder sortOrder) {
```
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/BuyClientDefault.java]: I find this overloading of the api a bit confusing.  I can see that tags can be used to filter if have/don't have a collection Id, but sortOrder is only used if collectionId is not null.
For clarity of the API I would prefer to see a bit of code duplication with two APIs, one with (page, collectionId, tags, sortOrder) where collectionId is mandatory and one with (page, tags) that gets all products filtered by  tags.
@bgulanowski @davidmuzi ?

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductService.java [hunk 8]
```diff
@@ -111,59 +112,63 @@
...
+     * Fetch the products filtered by specified collection and tags
      *
-     * @param page     the 1-based page index. The page size is set by {@link BuyClientBuilder#productPageSize} configuration.
-     * @param callback the {@link Callback} that will be used to indicate the response from the asynchronous network operation, not null
+     * @param page         the 1-based page index. The page size is set by {@link BuyClientBuilder#productPageSize} configuration.
+     * @param collectionId the collectionId that we want to fetch products for, can be null
+     * @param tags         set of tags which each product must contain, can be null
+     * @param sortOrder    the sort order of products for the specified collection,
+     *                     in case of {@code null} value {@link Collection.SortOrder#COLLECTION_DEFAULT} will be used as default,
+     *                     will be ignored if collection is not specified
```
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductService.java]: would be nice to link to collectionId here to make it clearer

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductService.java [hunk 9]
```diff
@@ -111,59 +112,63 @@
...
+     * Fetch the products filtered by specified collection and tags
      *
-     * @param page the 1-based page index. The page size is set by {@link BuyClientBuilder#productPageSize} configuration.
-     * @return cold observable that emits requested list of collections
+     * @param page         the 1-based page index. The page size is set by {@link BuyClientBuilder#productPageSize} configuration.
+     * @param collectionId the collectionId that we want to fetch products for, can be null
+     * @param tags         set of tags which each product must contain, can be null
+     * @param sortOrder    the sort order of products for the specified collection,
+     *                     in case of {@code null} value {@link Collection.SortOrder#COLLECTION_DEFAULT} will be used as default,
+     *                     will be ignored if collection is not specified
```
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductService.java]: same as above

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/BuyClientDefault.java (hunk 7 above)
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/BuyClientDefault.java]: right.  my comment is more about the current function being confusing, not adding about the number of parameters.

**Comment by krisorr** [comment]
LGTM
